File: packages/@emulators/aws/src/routes/inspector.ts
import type { RouteContext, InspectorTab } from "@emulators/core";
import { renderInspectorPage } from "@emulators/core";
import { getAwsStore } from "../store.js";
import { escapeXml } from "../helpers.js";

const SERVICE_LABEL = "AWS";

const TABS: InspectorTab[] = [
  { id: "s3", label: "S3", href: "/?tab=s3" },
  { id: "sqs", label: "SQS", href: "/?tab=sqs" },
  { id: "iam", label: "IAM", href: "/?tab=iam" },
];

export function inspectorRoutes(ctx: RouteContext): void {
  const { app, store } = ctx;
  const aws = () => getAwsStore(store);

  app.get("/", (c) => {
    const tab = c.req.query("tab") ?? "s3";
    const s3Store = aws();

    const buckets = s3Store.s3Buckets.all();
    const queues = s3Store.sqsQueues.all();
    const users = s3Store.iamUsers.all();
    const roles = s3Store.iamRoles.all();

    let contentHtml = "";

    if (tab === "s3") {
      const rows = buckets
        .map((b) => {
          const objects = s3Store.s3Objects.findBy("bucket_name", b.bucket_name);
          return `<tr>
            <td>${escapeXml(b.bucket_name)}</td>
            <td>${objects.length}</td>
            <td>${escapeXml(b.region)}</td>
            <td>${escapeXml(b.creation_date)}</td>
          </tr>`;
        })
        .join("\n");

      contentHtml = `
        <div class="inspector-section">
          <h2>S3 Buckets (${buckets.length})</h2>
          <table class="inspector-table">
            <thead><tr><th>Bucket</th><th>Objects</th><th>Region</th><th>Created</th></tr></thead>
            <tbody>${rows || `<tr><td colspan="4"><div class="inspector-empty">No buckets</div></td></tr>`}</tbody>
          </table>
        </div>`;

      for (const bucket of buckets) {
        const objects = s3Store.s3Objects.findBy("bucket_name", bucket.bucket_name);
        if (objects.length > 0) {
          const objRows = objects
            .map(
              (o) => `<tr>
              <td>${escapeXml(o.key)}</td>
              <td>${o.content_length}</td>
              <td>${escapeXml(o.content_type)}</td>
              <td>${escapeXml(o.last_modified)}</td>
            </tr>`,
            )
            .join("\n");
          contentHtml += `
            <div class="inspector-section">
              <h3>${escapeXml(bucket.bucket_name)} objects</h3>
              <table class="inspector-table">
                <thead><tr><th>Key</th><th>Size</th><th>Type</th><th>Last Modified</th></tr></thead>
                <tbody>${objRows}</tbody>
              </table>
            </div>`;
        }
      }
    } else if (tab === "sqs") {
      const rows = queues
        .map((q) => {
          const messages = s3Store.sqsMessages.findBy("queue_name", q.queue_name);
          return `<tr>
            <td>${escapeXml(q.queue_name)}</td>
            <td>${messages.length}</td>
            <td>${q.fifo ? "Yes" : "No"}</td>
            <td>${q.visibility_timeout}s</td>
          </tr>`;
        })
        .join("\n");

      contentHtml = `
        <div class="inspector-section">
          <h2>SQS Queues (${queues.length})</h2>
          <table class="inspector-table">
            <thead><tr><th>Queue</th><th>Messages</th><th>FIFO</th><th>Visibility Timeout</th></tr></thead>
            <tbody>${rows || `<tr><td colspan="4"><div class="inspector-empty">No queues</div></td></tr>`}</tbody>
          </table>
        </div>`;
    } else if (tab === "iam") {
      const userRows = users
        .map(
          (u) => `<tr>
          <td>${escapeXml(u.user_name)}</td>
          <td>${escapeXml(u.user_id)}</td>
          <td>${u.access_keys.length}</td>
          <td>${escapeXml(u.arn)}</td>
        </tr>`,
        )
        .join("\n");

      const roleRows = roles
        .map(
          (r) => `<tr>
          <td>${escapeXml(r.role_name)}</td>
          <td>${escapeXml(r.role_id)}</td>
          <td>${escapeXml(r.description)}</td>
          <td>${escapeXml(r.arn)}</td>
        </tr>`,
        )
        .join("\n");

      contentHtml = `
        <div class="inspector-section">
          <h2>IAM Users (${users.length})</h2>
          <table class="inspector-table">
            <thead><tr><th>User</th><th>User ID</th><th>Access Keys</th><th>ARN</th></tr></thead>
            <tbody>${userRows || `<tr><td colspan="4"><div class="inspector-empty">No users</div></td></tr>`}</tbody>
          </table>
        </div>
        <div class="inspector-section">
          <h2>IAM Roles (${roles.length})</h2>
          <table class="inspector-table">
            <thead><tr><th>Role</th><th>Role ID</th><th>Description</th><th>ARN</th></tr></thead>
            <tbody>${roleRows || `<tr><td colspan="4"><div class="inspector-empty">No roles</div></td></tr>`}</tbody>
          </table>
        </div>`;
    }

    return c.html(renderInspectorPage("Inspector", TABS, tab, contentHtml, SERVICE_LABEL));
  });
}

File: internal/services/aws/inspector.go
package aws

import (
	"fmt"
	"strings"

	corehttp "github.com/vercel-labs/emulate/internal/core/http"
	corestore "github.com/vercel-labs/emulate/internal/core/store"
	"github.com/vercel-labs/emulate/internal/core/ui"
)

var inspectorTabs = []ui.InspectorTab{
	{ID: "s3", Label: "S3", Href: "/_inspector?tab=s3"},
	{ID: "sqs", Label: "SQS", Href: "/_inspector?tab=sqs"},
	{ID: "iam", Label: "IAM", Href: "/_inspector?tab=iam"},
	{ID: "logs", Label: "Logs", Href: "/_inspector?tab=logs"},
}

func (s *Service) handleInspector(c *corehttp.Context) {
	tab := c.Query("tab")
	if tab != "s3" && tab != "sqs" && tab != "iam" && tab != "logs" {
		tab = "s3"
	}
	tabs := make([]ui.InspectorTab, len(inspectorTabs))
	for index, item := range inspectorTabs {
		tabs[index] = item
		tabs[index].Active = item.ID == tab
	}

	var body string
	switch tab {
	case "sqs":
		body = s.renderSQSInspector()
	case "iam":
		body = s.renderIAMInspector()
	case "logs":
		body = s.renderLogsInspector()
	default:
		body = s.renderS3Inspector()
	}

	c.HTML(200, ui.RenderInspectorPage("Inspector", tabs, tab, body, ui.PageOptions{Service: "AWS"}))
}

func (s *Service) renderS3Inspector() string {
	buckets := s.store.S3Buckets.All()
	var rows strings.Builder
	for _, bucket := range buckets {
		name := stringField(bucket, "bucket_name")
		objects := s.store.S3Objects.FindBy("bucket_name", name)
		rows.WriteString(`<tr><td>`)
		rows.WriteString(ui.EscapeHTML(name))
		rows.WriteString(`</td><td>`)
		rows.WriteString(fmt.Sprint(len(objects)))
		rows.WriteString(`</td><td>`)
		rows.WriteString(ui.EscapeHTML(stringField(bucket, "region")))
		rows.WriteString(`</td><td>`)
		rows.WriteString(ui.EscapeHTML(stringField(bucket, "creation_date")))
		rows.WriteString(`</td></tr>`)
	}
	return `<div class="inspector-section">
  <h2>S3 Buckets (` + fmt.Sprint(len(buckets)) + `)</h2>
  <table class="inspector-table">
    <thead><tr><th>Bucket</th><th>Objects</th><th>Region</th><th>Created</th></tr></thead>
    <tbody>` + rowsOrEmpty(rows.String(), 4, "No buckets") + `</tbody>
  </table>
</div>`
}

func (s *Service) renderSQSInspector() string {
	queues := s.store.SQSQueues.All()
	var rows strings.Builder
	for _, queue := range queues {
		name := stringField(queue, "queue_name")
		messages := s.store.SQSMessages.FindBy("queue_name", name)
		rows.WriteString(`<tr><td>`)
		rows.WriteString(ui.EscapeHTML(name))
		rows.WriteString(`</td><td>`)
		rows.WriteString(fmt.Sprint(len(messages)))
		rows.WriteString(`</td><td>`)
		rows.WriteString(yesNo(boolField(queue, "fifo")))
		rows.WriteString(`</td><td>`)
		rows.WriteString(ui.EscapeHTML(stringField(queue, "visibility_timeout")))
		rows.WriteString(`</td></tr>`)
	}
	return `<div class="inspector-section">
  <h2>SQS Queues (` + fmt.Sprint(len(queues)) + `)</h2>
  <table class="inspector-table">
    <thead><tr><th>Queue</th><th>Messages</th><th>FIFO</th><th>Visibility Timeout</th></tr></thead>
    <tbody>` + rowsOrEmpty(rows.String(), 4, "No queues") + `</tbody>
  </table>
</div>`
}

func (s *Service) renderIAMInspector() string {
	users := s.store.IAMUsers.All()
	roles := s.store.IAMRoles.All()
	var userRows strings.Builder
	for _, user := range users {
		userRows.WriteString(`<tr><td>`)
		userRows.WriteString(ui.EscapeHTML(stringField(user, "user_name")))
		userRows.WriteString(`</td><td>`)
		userRows.WriteString(ui.EscapeHTML(stringField(user, "user_id")))
		userRows.WriteString(`</td><td>`)
		userRows.WriteString(fmt.Sprint(accessKeyCount(user)))
		userRows.WriteString(`</td><td>`)
		userRows.WriteString(ui.EscapeHTML(stringField(user, "arn")))
		userRows.WriteString(`</td></tr>`)
	}
	var roleRows strings.Builder
	for _, role := range roles {
		roleRows.WriteString(`<tr><td>`)
		roleRows.WriteString(ui.EscapeHTML(stringField(role, "role_name")))
		roleRows.WriteString(`</td><td>`)
		roleRows.WriteString(ui.EscapeHTML(stringField(role, "role_id")))
		roleRows.WriteString(`</td><td>`)
		roleRows.WriteString(ui.EscapeHTML(stringField(role, "description")))
		roleRows.WriteString(`</td><td>`)
		roleRows.WriteString(ui.EscapeHTML(stringField(role, "arn")))
		roleRows.WriteString(`</td></tr>`)
	}
	return `<div class="inspector-section">
  <h2>IAM Users (` + fmt.Sprint(len(users)) + `)</h2>
  <table class="inspector-table">
    <thead><tr><th>User</th><th>User ID</th><th>Access Keys</th><th>ARN</th></tr></thead>
    <tbody>` + rowsOrEmpty(userRows.String(), 4, "No users") + `</tbody>
  </table>
</div>
<div class="inspector-section">
  <h2>IAM Roles (` + fmt.Sprint(len(roles)) + `)</h2>
  <table class="inspector-table">
    <thead><tr><th>Role</th><th>Role ID</th><th>Description</th><th>ARN</th></tr></thead>
    <tbody>` + rowsOrEmpty(roleRows.String(), 4, "No roles") + `</tbody>
  </table>
</div>`
}

func (s *Service) renderLogsInspector() string {
	groups := s.store.LogGroups.All()
	var groupRows strings.Builder
	for _, group := range groups {
		name := stringField(group, "log_group_name")
		streamCount := 0
		for _, stream := range s.store.LogStreams.FindBy("log_group_name", name) {
			if stringField(stream, "account_id") == stringField(group, "account_id") && stringField(stream, "region") == stringField(group, "region") {
				streamCount++
			}
		}
		eventCount := 0
		for _, event := range s.store.LogEvents.FindBy("log_group_name", name) {
			if stringField(event, "account_id") == stringField(group, "account_id") && stringField(event, "region") == stringField(group, "region") {
				eventCount++
			}
		}
		groupRows.WriteString(`<tr><td>`)
		groupRows.WriteString(ui.EscapeHTML(name))
		groupRows.WriteString(`</td><td>`)
		groupRows.WriteString(fmt.Sprint(streamCount))
		groupRows.WriteString(`</td><td>`)
		groupRows.WriteString(fmt.Sprint(eventCount))
		groupRows.WriteString(`</td><td>`)
		groupRows.WriteString(ui.EscapeHTML(stringField(group, "retention_in_days")))
		groupRows.WriteString(`</td><td>`)
		groupRows.WriteString(ui.EscapeHTML(stringField(group, "arn")))
		groupRows.WriteString(`</td></tr>`)
	}
	return `<div class="inspector-section">
  <h2>CloudWatch Log Groups (` + fmt.Sprint(len(groups)) + `)</h2>
  <table class="inspector-table">
    <thead><tr><th>Log Group</th><th>Streams</th><th>Events</th><th>Retention Days</th><th>ARN</th></tr></thead>
    <tbody>` + rowsOrEmpty(groupRows.String(), 5, "No log groups") + `</tbody>
  </table>
</div>`
}

func rowsOrEmpty(rows string, columns int, label string) string {
	if rows != "" {
		return rows
	}
	return `<tr><td colspan="` + fmt.Sprint(columns) + `"><div class="inspector-empty">` + ui.EscapeHTML(label) + `</div></td></tr>`
}

func stringField(record corestore.Record, name string) string {
	switch value := record[name].(type) {
	case string:
		return value
	case int:
		return fmt.Sprint(value)
	case int64:
		return fmt.Sprint(value)
	case float64:
		return fmt.Sprint(value)
	default:
		return ""
	}
}

func boolField(record corestore.Record, name string) bool {
	value, _ := record[name].(bool)
	return value
}

func yesNo(value bool) string {
	if value {
		return "Yes"
	}
	return "No"
}

func accessKeyCount(record corestore.Record) int {
	switch value := record["access_keys"].(type) {
	case []corestore.Record:
		return len(value)
	case []any:
		return len(value)
	case []map[string]any:
		return len(value)
	default:
		return 0
	}
}
